.circleci/config.yml

```diff
@@ -0,0 +1,135 @@
+    docker:
+      - image: cimg/python:3.8.13
+    steps:
+      - checkout
+      - run:
+          name: "Install requirements"
+          command: |
+            pip install --upgrade pip
+            pip install -r requirements.txt
+            pip install -r requirements-dev.txt
+      - run:
+          name: "Code Quality Check"
+          command: |
+            black dynamicio
+            black tests
+            flake8 --verbose dynamicio
+            flake8 --verbose tests
+            pylint -v dynamicio
+            pylint -v tests
+            yamllint -v dynamicio
+            yamllint -v tests
+      - run:
+          name: "Run Python Tests"
+          command: |
+            pytest --cache-clear --cov=dynamicio tests/
+  test_python_3_9:
+    working_directory: ~/project
+    docker:
+      - image: cimg/python:3.9.13
+    steps:
+      - checkout
+      - run:
+          name: "Install requirements"
+          command: |
+            pip install --upgrade pip
+            pip install -r requirements.txt
+            pip install -r requirements-dev.txt
+      - run:
+          name: "Code Quality Check"
+          command: |
+            black dynamicio
+            black tests
+            flake8 --verbose dynamicio
+            flake8 --verbose tests
+            pylint -v dynamicio
+            pylint -v tests
+            yamllint -v dynamicio
+            yamllint -v tests
+      - run:
+          name: "Run Python Tests"
+          command: |
+            pytest --cache-clear --cov=dynamicio tests/
+  test_python_3_10:
+    working_directory: ~/project
+    docker:
+      - image: cimg/python:3.10.4
+    steps:
+      - checkout
+      - run:
+          name: "Install requirements"
+          command: |
+            pip install --upgrade pip
+            pip install -r requirements.txt
+            pip install -r requirements-dev.txt
+      - run:
+          name: "Code Quality Check"
+          command: |
+            black dynamicio
+            black tests
+            flake8 --verbose dynamicio
+            flake8 --verbose tests
+            pylint -v dynamicio
+            pylint -v tests
+            yamllint -v dynamicio
+            yamllint -v tests
+      - run:
+          name: "Run Python Tests"
+          command: |
+            pytest --cache-clear --cov=dynamicio tests/
+
+
+  build_and_publish:
+    working_directory: ~/project
+    docker:
+      - image: cimg/python:3.8.4
+    steps:
+      - checkout
+      - run:
+          name: "Install build requirements"
+          command: |
+            pip install --upgrade pip
+            pip install -r requirements-build.txt
+      - run:
+          name: "Build package"
+          command: |
+            BUILD_VERSION=${CIRCLE_TAG:1} python setup.py sdist bdist_wheel
+      - run:
+          name: "Publish"
+          command: |
+            pip install --upgrade twine
+            twine check dist/*
+            twine upload dist/*
```

Comment:
Just to try and kick the tyres - how does the twine upload work? what credentials is it using?

Reply:
Yes, so, locally I use a `.pypirc` file which contains my credentials.

BUt we can pass arguments to twine to get it to work:
```shell
optional arguments:
  -h, --help            show this help message and exit
  -r REPOSITORY, --repository REPOSITORY
                        The repository (package index) to upload the package to. Should be a section in the config file (default: pypi). (Can also be set via TWINE_REPOSITORY environment variable.)
  --repository-url REPOSITORY_URL
                        The repository (package index) URL to upload the package to. This overrides --repository. (Can also be set via TWINE_REPOSITORY_URL environment variable.)
  -s, --sign            Sign files to upload using GPG.
  --sign-with SIGN_WITH
                        GPG program used to sign uploads (default: gpg).
  -i IDENTITY, --identity IDENTITY
                        GPG identity used to sign files.
  -u USERNAME, --username USERNAME
                        The username to authenticate to the repository (package index) as. (Can also be set via TWINE_USERNAME environment variable.)
  -p PASSWORD, --password PASSWORD
                        The password to authenticate to the repository (package index) with. (Can also be set via TWINE_PASSWORD environment variable.)
  --non-interactive     Do not interactively prompt for username/password if the required credentials are missing. (Can also be set via TWINE_NON_INTERACTIVE environment variable.)
  -c COMMENT, --comment COMMENT
                        The comment to include with the distribution file.
  --config-file CONFIG_FILE
                        The .pypirc config file to use.
  --skip-existing       Continue uploading files if one already exists. (Only valid when uploading to PyPI. Other implementations may not support this.)
  --cert path           Path to alternate CA bundle (can also be set via TWINE_CERT environment variable).
  --client-cert path    Path to SSL client certificate, a single file containing the private key and the certificate in PEM format.
  --verbose             Show verbose output.
  --disable-progress-bar
                        Disable the progress bar.
```

Reply:
Actually, we don't even need to do that. See here: https://twine.readthedocs.io/en/stable/#configuration

Reply:
All, done. I have set those up in circle-ci

Reply:
cool.